src/jit/lowerxarch.cpp

```diff
@@ -2113,10 +2145,44 @@ void Lowering::ContainCheckBinary(GenTreeOp* node)
         // If this binary op neither has contained operands, nor is a
         // Read-Modify-Write (RMW) operation, we can mark its operands
         // as reg optional.
-        SetRegOptionalForBinOp(node);
+
+        isSafeToContainOp1 = isSafeToContainOp1 && IsSafeToContainMem(node, op1);
+        isSafeToContainOp2 = isSafeToContainOp2 && IsSafeToContainMem(node, op2);
+
+        SetRegOptionalForBinOp(node, isSafeToContainOp1, isSafeToContainOp2);
     }
 }
 
+//------------------------------------------------------------------------
+// IsSafelyContainableMemoryOp: Determine if this is a containable memory operand
+//                              that can be safely contained
+//
+// Arguments:
+//    parentNode        - Parent node of the operand
+//    childNode         - Child node to check for safe containment
+//    isSafeToContainOp - Out-parameter that is set to false if we determined that the operand
+//                        is not safe to contain, true otherwise.
+//
+// Return Value:
+//    Returns true if the operand can be safely contained.
+//
+bool Lowering::IsSafelyContainableMemoryOp(GenTree* parentNode, GenTree* childNode, bool* isSafeToContainOp)
```

Comment:
I think this method is confusing. First, the name is too similar to `IsSafeToContainMem`. Second, the difference between the out parameter and the result is only for the case where `childNode` is not a memory op - I think that's inappropriate for a method specifically focused on memory. I think it is better, and articulates the cases more clearly, for the caller to do that checking, and then to call `IsSafeToContainMem` directly, even if it means there are more conditions for the caller to check.

Reply:
I did it the way you are requesting originally and didn't like the code in the callers. I'll re-do and push that version to show what I got.

Reply:
@CarolEidt I pushed the changes. Do you like this version better?

Reply:
Yes - I think it's much easier to understand this way. Thanks!